File: core/x/javaapi/api/src/main/java/viaduct/java/api/types/OffsetLimit.java
package viaduct.java.api.types;

/**
 * Result of converting {@link ConnectionArguments} to an offset/limit pair.
 *
 * <p>This is the bridge between GraphQL pagination arguments ({@code first}, {@code after}, {@code
 * last}, {@code before}) and the offset-based slice that a backend query requires. Produced by
 * {@link ConnectionArguments#toOffsetLimit()} and consumed by {@link
 * viaduct.java.api.internal.ConnectionBuilder} internally.
 *
 * <p>Java equivalent of Kotlin's {@code viaduct.api.types.OffsetLimit}.
 *
 * @param offset index of the first item to fetch; a negative value signals {@link
 *     viaduct.java.api.internal.ConnectionBuilder#fromList} to resolve from the tail of the dataset
 *     (e.g. {@code -1} is the last item, {@code -10} means the 10th from the end)
 * @param limit maximum number of items to fetch
 */
public record OffsetLimit(int offset, int limit) {}

File: core/tenant/api/src/main/kotlin/viaduct/api/types/OffsetLimit.kt
package viaduct.api.types

import viaduct.apiannotations.ExperimentalApi

/**
 * Result of converting [viaduct.api.types.ConnectionArguments] to an offset/limit pair.
 *
 * This is the bridge between GraphQL pagination arguments (`first`, `after`, `last`, `before`)
 * and the offset-based slice that a backend query requires. Produced by
 * [viaduct.api.types.ConnectionArguments.toOffsetLimit] and consumed by
 * [viaduct.api.internal.ConnectionBuilder] internally.
 *
 * @property offset Index of the first item to fetch. A negative value signals
 *   [viaduct.api.internal.ConnectionBuilder.fromList] to resolve from the tail of the dataset
 *  (e.g. `-1` is the last item, `-10` means the 10th from the end).
 * @property limit Maximum number of items to fetch.
 */
@ExperimentalApi
data class OffsetLimit(
    val offset: Int,
    val limit: Int,
)
